blockservice/blockservice.go

```diff
@@ -54,13 +62,27 @@ func (s *BlockService) AddBlock(b blocks.Block) (key.Key, error) {
 }
 
 func (s *BlockService) AddBlocks(bs []blocks.Block) ([]key.Key, error) {
-	err := s.Blockstore.PutMany(bs)
+	var toput []blocks.Block
+	for _, b := range bs {
+		has, err := s.Blockstore.Has(b.Key())
+		if err != nil {
+			return nil, err
+		}
+
+		if has {
+			continue
+		}
+
+		toput = append(toput, b)
+	}
+
+	err := s.Blockstore.PutMany(toput)
 	if err != nil {
 		return nil, err
 	}
 
 	var ks []key.Key
-	for _, b := range bs {
+	for _, b := range toput {
 		if err := s.Exchange.HasBlock(b); err != nil {
```

Comment:
@whyrusleeping, this change modifies the return value so that only the keys that don't already exist are returned.  Was this intentional?